File: 3. Behavioral Design Patterns/Command/Kotlin/CommandTemplate.kt
// Command Pattern – Kotlin Template

interface Command { fun execute(); fun undo() }

class Receiver { fun action(){ /* do something */ }; fun undoAction(){ /* undo something */ } }

class ConcreteCommand(private val r: Receiver) : Command {
    override fun execute() = r.action()
    override fun undo() = r.undoAction()
}

class Invoker { private var cmd: Command? = null; fun set(c: Command){ cmd = c }; fun run(){ cmd?.execute() }; fun undo(){ cmd?.undo() } }

fun main(){ val inv = Invoker(); inv.set(ConcreteCommand(Receiver())); inv.run(); inv.undo() }

File: 3. Behavioral Design Patterns/Command/Go/CommandTemplate.go
package main

// Command Pattern – Go Template

type Command interface{ Execute(); Undo() }

type Receiver struct{}
func (Receiver) Action(){ /* do something */ }
func (Receiver) UndoAction(){ /* undo something */ }

type ConcreteCommand struct{ r Receiver }
func (c ConcreteCommand) Execute(){ c.r.Action() }
func (c ConcreteCommand) Undo(){ c.r.UndoAction() }

type Invoker struct{ cmd Command }
func (i *Invoker) Set(c Command){ i.cmd = c }
func (i Invoker) Run(){ if i.cmd != nil { i.cmd.Execute() } }
func (i Invoker) Undo(){ if i.cmd != nil { i.cmd.Undo() } }

func main(){ inv := Invoker{}; inv.Set(ConcreteCommand{Receiver{}}); inv.Run(); inv.Undo() }
